core/src/main/java/org/apache/iceberg/DataFiles.java

```diff
@@ -43,6 +43,7 @@ static PartitionData newPartitionData(PartitionSpec spec) {
   }
 
   static PartitionData copyPartitionData(PartitionSpec spec, StructLike partitionData, PartitionData reuse) {
+    Preconditions.checkArgument(!spec.isUnpartitioned(), "Can't copy partition data to a unpartitioned table");
```

Comment:
The underlying method explicitly check for `isUnpartitioned()` in case this method is called outside the builder. (`FileMetadata.java`)